### ISSUE TITLE
[java] LiteralsFirstInComparisonsRule not applied on constant

### ISSUE THREAD

**Issue by Pankraz76** [pmd/pmd#5590]
**Affects PMD Version:** `pmd:3.26.0`

**Rule: LiteralsFirstInComparisonsRule**

Please provide the rule name and a link to the rule documentation:

https://pmd.github.io/pmd/pmd_rules_java_bestpractices.html#literalsfirstincomparisons

**Description:**

It is not found on the constant.  
Is this indented correctly?  
After briefly reviewing the code, I don't see any filtering, so this might be accidental.  
A reference is always a potential `NPE`, in my opinion.  
Therefore, the raw string literal should always come first.

**Code Sample demonstrating the issue:**

```java
    static class Foo {
        private static final String FOO = null;

        public static void bar(String _null) {
            String _null2 = null;
            FOO.equals("RAW"); // not reported -> false-negative?
            _null.equals("RAW"); // is reported
            _null2.equals("RAW"); // is reported
        }
    }
```

<img width="881" alt="Image" src="https://github.com/user-attachments/assets/7a321a24-0201-4bb2-abc6-d0f42fff4598" />

**Comment by pankratz76** [comment]
https://github.com/openrewrite/rewrite-static-analysis/pull/473#issuecomment-2717705637

**Comment by adangel** [comment]
The rule made the assumption, that if you declare a constant string (that is a final static field of type String), it would never be initialized with null - and hence simply excluded all final static String fields.

The exclusion happens here: `!isContantString(qualifier)`:
https://github.com/pmd/pmd/blob/0a351423eb0a7f006a9c33f4badd630f139763f9/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/bestpractices/LiteralsFirstInComparisonsRule.java#L78

And "isContantString" is either a String literal or a final static field of type String. Everything else (local vars, non-final/non-static fields, method params) are not considered constant strings.

Note: initializing a string constant with null and using it in the way you described, should be pretty obvious at runtime - unless you never execute/test your code, you'll quickly see this problem as a NPE - regardless whether you use PMD or not.

@oowekyala is reworking this rule to not exclude final static fields anymore, but rather check, whether the compile-time constant of the qualifier is a String (and null isn't).

**Comment by pankratz76** [comment]
thanks for the clarification. The assumption is likely gonna work in most cases.
